### PR TITLE
Allow TypeScript syntax in arrow function parameters and return types

### PR DESCRIPTION
**Pull request by Dominator008**
[Review on Reviewable](https://reviewable.io/reviews/google/closure-compiler/1378)

### CODE REVIEW

**Comment by Dominator008** [comment]
Not quite working yet :(

File: src/com/google/javascript/jscomp/parsing/parser/Parser.java [hunk 1]
```diff
@@ -2620,16 +2620,29 @@ private boolean peekAssignmentExpression() {
...
+  }
+
+  private ParseTree parseAssignment(Expression expressionIn, boolean mayHaveColonType) {
     if (peek(TokenType.YIELD) && inGeneratorContext()) {
       return parseYield(expressionIn);
     }
 
     SourcePosition start = getTreeStartLocation();
-    // TODO(blickly): Allow TypeScript syntax in arrow function parameters
+    GenericTypeListTree generics = maybeParseGenericTypes();
```
Comment (blickly) [inline src/com/google/javascript/jscomp/parsing/parser/Parser.java]: Isn't this allowing parsing of generics outside arrow functions?

Let's add some tests that generics in random places are forbidden.
e.g.
var x  = <X>(5 + 7);

Comment (Dominator008) [inline src/com/google/javascript/jscomp/parsing/parser/Parser.java]: Done.

Comment (blickly) [inline src/com/google/javascript/jscomp/parsing/parser/Parser.java]: I think there may be a similar issue for COLON below.

Comment (Dominator008) [inline src/com/google/javascript/jscomp/parsing/parser/Parser.java]: Done.

**Comment by blickly** [comment]
One more test case:

```javascript
function f(cond) {
  return cond ?
        (v = 7, f = ()=>v) :
        n=>2*n;
}
```

Can we check that this still parses correctly as an ?: (since this is valid non TS ES6 code).

**Comment by Dominator008** [comment]
@blickly Done adding that test case, which is already parsed correctly.

**Comment by Dominator008** [comment]
@blickly PTAL?

**Comment by blickly** [comment]
Looks good. I'll pull it in.

**Comment by Dominator008** [comment]
@blickly Do you have a test case for the parse error in `switch` statements?

**Comment by blickly** [comment]
I think it was parsing the COLON of the case as a type expression.
```javascript
switch (EXPR1) {
   case EXPR2:
       break;
}
```
The particular `EXPR2` that was broken was a comma expression, I think, but since `EXPR2` can be an arbitrary expression, we probably need better test coverage of case expressions.

**Comment by nicks** [comment]
lol, sorry, github's trying to be too smart

**Comment by MatrixFrog** [comment]
Can you try sending this internally if you still want this?

**Pull request closed without merging**